### ISSUE TITLE
Add support to multiple formatters or access to configuration options from formatter to create a tee-like formatter

### ISSUE THREAD

**Issue by alejandroclaro** [eslint/eslint#13244]
**The version of ESLint you are using.**
6.6.0

**The problem you want to solve.**
Generation of different outputs using formatters without running eslint cli several times. A specific use case we have is the generation of a checkstyle report on our CI server, but also the inspection of results the standard output in both CI server and development environment.

**Your take on the correct solution to problem.**
I envision two simple solutions that require the support of eslint cli:

1. Support of multiple formats and their respective configuration parameters such as their output file. This support could be limited to when such configuration is provided from a configuration file (e.g. eslintrc.js), and when provided by arguments to cli, it's limited to a single formatter.

2. Allow the configuration or part of it (some section dedicated to the formatters) to reach the formatter so that it is possible to build more interesting formatters such as a formatter that demux the result to multiple fomatters (e.g. compose multiple fomatters).

This request is similar to some other requests that have been made. For example #10708. However, I think the proposed solutions are different and broad enough to allow the implementation of other interesting use cases based on this.

Also, there is an existing project that tries to provide a similar solution ([eslint-output](https://github.com/lwhiteley/eslint-output)) based on the CLIEngine, but it doesn't play nicely with monorepos, requires additional configuration files, and it's a different cli tool which we do not consider something good.

**Comment by anikethsaha** [comment]
> A specific use case we have is the generation of a checkstyle report on our CI server, but also the inspection of results the standard output in both CI server and development environment.

I think having a custom formatter and reading the environment variable and run the specific formatter conditionally can be a solution for this case. 

 about your second solution 

> Allow the configuration or part of it (some section dedicated to the formatters) to reach the formatter so that it is possible to build more interesting formatters such as a formatter that demux the result to multiple fomatters (e.g. compose multiple fomatters).

is it something like passing data from the config file to the formatter ?

**Comment by alejandroclaro** [comment]
Hi @anikethsaha,

> I think having a custom formatter and reading the environment variable and run the specific formatter conditionally can be a solution for this case.

It is not very elegant, but it could work. However, there are still some problems with this:

1. We have many packages in a monorepo. We manage the linter configuration exporting the configuration from a package with preset configurations for the building and analysis tools. One thing we don't want to do is to write the invocation of eslint with the formatter argument in every place (e.g. NPM script on each package). For this, it will be very nice to have the capability to configure the formatter in a preset configuration file.

2.  Also it's a problem with how the formatters lookup works. It assumes that the package only exports the (one) formatter, and we use other tools besides eslint. It a burden have to develop a package for each formatter, or any other extension, for each tool.

eslint is very good at letting client to configure how input is managed and processed from configuration files, but not so good providing extension points to manage the output.

> is it something like passing data from the config file to the formatter ?

Yes. I tried to add a section in the eslint configuration to manage the formatter using the CLIEngine API. Something like this:

``` javascript
module.exports = {
  'formatters': [
    {
      'name': 'stylish',
      'output': 'eslint.log'
    },
    {
      'name': 'checkstyle',
      'output': './static-analysis/code.checkstyle.xml'
    }
  ],
  // rest of the eslint configuration
};
````

However, of course, it fails as it is not a valid field in the configuration. It will be nice to be able to configure formatters parameter this way instead of having to inject environment variables, and writing a bash script to do it plus invoking eslint, instead of just letting the configuration be a configuration in a configuration file.

My point is that it is very difficult to get the formatters to get configuration parameters and also reuse the configuration for them. It is a big limitation that is only available through the CLI parameters and no parameter configuration for them exists (the env variables is more like a hack).

I was trying to build my own CLI using the CLIEngine class but many important pieces are in the cli.js module and eslint.js application. So I have been copy-pasting a lot of things to make it work the way we want, which is very simple. We want to be able to control the output format from the preset configuration and be able to produce as many output formats as we need without running the tool many times.
